File: app/src/main/java/de/tomcory/heimdall/net/flow/cache/DnsCache.java
package de.tomcory.heimdall.net.flow.cache;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.LinkedHashMap;

public class DnsCache {
    private static final int MAX_SIZE = 128;

    //the singleton
    private static final DnsCache cache = new DnsCache();

    private final LinkedHashMap<String, String> hosts = new LinkedHashMap<>(MAX_SIZE) {
        @Override
        protected boolean removeEldestEntry(Entry<String, String> eldest) {
            return size() >= MAX_SIZE;
        }
    };

    @Nullable
    public static String findHost(@NonNull String ipAddress) {
        return cache.hosts.get(ipAddress);
    }

    public static void addHost(@NonNull String ipAddress, @NonNull String hostname) {
        cache.hosts.put(ipAddress, hostname);
    }
}

File: core/vpn/src/main/java/de/tomcory/heimdall/core/vpn/cache/DnsCache.kt
package de.tomcory.heimdall.core.vpn.cache

import java.util.LinkedHashMap

class DnsCache {
    private val hosts: LinkedHashMap<String?, String?> = object : LinkedHashMap<String?, String?>(MAX_SIZE) {
        override fun removeEldestEntry(eldest: Map.Entry<String?, String?>): Boolean {
            return size >= MAX_SIZE
        }
    }

    companion object {
        private const val MAX_SIZE = 128

        //the singleton
        private val cache = DnsCache()
        @JvmStatic
        fun findHost(ipAddress: String): String? {
            return cache.hosts[ipAddress]
        }

        fun addHost(ipAddress: String, hostname: String) {
            cache.hosts[ipAddress] = hostname
        }
    }
}